internal/streamingcoord/server/balancer/policy/vchannelfair/expected_layout.go

```diff
@@ -0,0 +1,173 @@
+	node, ok := p.CurrentLayout.AllNodesInfo[serverID]
+	if !ok {
+		panic("node info not found")
+	}
+
+	// assign to the node that already has pchannel at highest priority.
+	p.Assignments[channelID] = node
+	p.Nodes[node.ServerID].AssignedChannels[channelID] = struct{}{}
+	p.Nodes[node.ServerID].AssignedVChannelCount += len(stats.VChannels)
+	p.updateNodeScore(node.ServerID)
+}
+
+// Unassign will unassign the channel from the node.
+func (p *expectedLayoutForVChannelFairPolicy) Unassign(channelID types.ChannelID) {
+	node, ok := p.Assignments[channelID]
+	if !ok {
+		panic("channel is not assigned")
+	}
+	delete(p.Assignments, channelID)
+	delete(p.Nodes[node.ServerID].AssignedChannels, channelID)
+	p.Nodes[node.ServerID].AssignedVChannelCount -= len(p.CurrentLayout.Channels[channelID].VChannels)
+	p.updateNodeScore(node.ServerID)
+}
+
+// updateNodeScore will update the score for the node.
+func (p *expectedLayoutForVChannelFairPolicy) updateNodeScore(serverID int64) {
+	newUnbalancedScore := p.currentCost(p.Nodes[serverID])
+	diff := newUnbalancedScore - p.Nodes[serverID].UnbalancedScore
+	p.GlobalUnbalancedScore += diff
+	p.Nodes[serverID].UnbalancedScore = newUnbalancedScore
+}
+
+// updateAllScore will update the score for all nodes and the global unbalanced score.
+func (p *expectedLayoutForVChannelFairPolicy) updateAllScore() {
+	// update the global unbalanced score.
+	p.GlobalUnbalancedScore = 0
+	for _, node := range p.Nodes {
+		node.UnbalancedScore = p.currentCost(node)
+		p.GlobalUnbalancedScore += node.UnbalancedScore
+	}
+}
+
+// currentCost will calculate the cost of the channel on the node.
+func (p *expectedLayoutForVChannelFairPolicy) currentCost(nodeInfo *streamingNodeInfo) float64 {
+	cost := float64(0.0)
+	if p.AveragePChannelPerNode != 0 {
+		pDiff := (float64(len(nodeInfo.AssignedChannels)) - p.AveragePChannelPerNode) / p.AveragePChannelPerNode
+		cost += p.Config.PChannelWeight * (pDiff * pDiff)
+	}
+	if p.AverageVChannelPerNode != 0 {
+		vDiff := (float64(nodeInfo.AssignedVChannelCount) - p.AverageVChannelPerNode) / p.AverageVChannelPerNode
+		cost += p.Config.VChannelWeight * (vDiff * vDiff)
+	}
+	assigned := lo.Keys(nodeInfo.AssignedChannels)
+	for i := 0; i < len(assigned); i++ {
+		for j := i + 1; j < len(assigned); j++ {
+			cost += p.Config.AntiAffinityWeight * (1 - p.PChannelAffinity.MustGetAffinity(assigned[i], assigned[j]))
+		}
+	}
+	return cost
+}
+
+// FindTheLeastUnbalanceScoreIncrementChannel will find the channel that increases the least score.
+func (p *expectedLayoutForVChannelFairPolicy) FindTheLeastUnbalanceScoreIncrementChannel() types.ChannelID {
+	var targetChannelID types.ChannelID
+	minScore := math.MaxFloat64
+	for channelID := range p.Assignments {
+		serverID := p.Assignments[channelID].ServerID
+		p.Unassign(channelID)
+		currentScore := p.GlobalUnbalancedScore
+		if currentScore < minScore {
+			minScore = currentScore
+			targetChannelID = channelID
+		}
+		p.Assign(channelID, serverID)
```

Comment:
seems a bit heavy impl, if there are n channels, to find the best one, we need to call 2n `updateNodeScore`